### PR TITLE
Log filtering framework client

### PR DESCRIPTION
**Pull request by tejaswiagarwal**
There was no mechanism to filter logs in the client. This adds a new framework in client using ContextExtractor to filter logs.

### CODE REVIEW

File: runtime/tchannel_client.go [hunk 1]
```diff
@@ -154,7 +156,16 @@ func (c *TChannelClient) Call(
...
 	}
-
+	var logFields []zap.Field
+	if extractor != nil {
+		headers := map[string]string{}
+		for k, v := range reqHeaders {
+			headers[k] = v
+		}
+		ctx = WithEndpointRequestHeadersField(ctx, headers)
+		logFields = append(logFields, extractor.ExtractLogFields(ctx)...)
```
[the remark this answers is not in the document]
Comment (jacobgreenleaf) [reply runtime/tchannel_client.go]: Log fields on context are required if e.g. the endpoint adds log fields using `WithLogFields(context.Context, ...zap.Field) context.Context` and wants the platform logs to have those fields.

[the remark this answers is not in the document]
Comment (ChuntaoLu) [reply runtime/tchannel_client.go]: Right, my point is client, being the bottom layer of the stack, will not propagate context further. We don't have to put log fields on the context if we extract log fields at log site.

File: runtime/tchannel_interfaces.go [hunk 2]
```diff
@@ -36,7 +36,7 @@ type RWTStruct interface {
 type TChannelCaller interface {
 	// Call should be passed the method to call, headers and the request/response thriftrw structs.
 	// The arguments returned are (whether there was an application error, unexpected error)
-	Call(ctx context.Context, service, method string, reqHeaders map[string]string, req, resp RWTStruct) (success bool, respHeaders map[string]string, err error)
+	Call(ctx context.Context, service, method string, reqHeaders map[string]string, req, resp RWTStruct, extractor ContextExtractor) (success bool, respHeaders map[string]string, err error)
```
[the remark this answers is not in the document]
Comment (tejaswiagarwal) [reply runtime/tchannel_interfaces.go]: Done!

File: test/endpoints/tchannel/baz/baz_simpleservice_method_call_test.go [hunk 3]
```diff
@@ -158,8 +158,6 @@ func TestCallTChannelSuccessfulRequestOKResponse(t *testing.T) {
 		"hostname",
 		"pid",
 		"timestamp-finished",
-		"Client-Req-Header-x-request-uuid",
```
[the remark this answers is not in the document]
Comment (tejaswiagarwal) [reply test/endpoints/tchannel/baz/baz_simpleservice_method_call_test.go]: The code to add "Client-Req-Header" was removed from here: https://github.com/uber/zanzibar/pull/608/files/4728e886495b2d2fa9f1d0a92a3d83b0c92947a9#diff-63b43474c6820d3db553746e13b0e82fL92

Hence, these logs will not be generated anymore.

File: runtime/tchannel_client.go [hunk 4]
```diff
@@ -69,8 +69,9 @@ type TChannelClientOption struct {
 
 // TChannelClient implements TChannelCaller and makes outgoing Thrift calls.
 type TChannelClient struct {
-	ClientID string
-	Loggers  map[string]*zap.Logger
+	ClientID  string
+	Extractor ContextExtractor
```
[the remark this answers is not in the document]
Comment (tejaswiagarwal) [reply runtime/tchannel_client.go]: Done

File: runtime/tchannel_client.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (tejaswiagarwal) [reply runtime/tchannel_client.go]: Done, these have been moved to tchannel_outbound_call.go

File: runtime/tchannel_client.go [hunk 5]
```diff
@@ -146,6 +147,7 @@ func (c *TChannelClient) Call(
 	thriftService, methodName string,
 	reqHeaders map[string]string,
 	req, resp RWTStruct,
+	extractor ContextExtractor,
```
[the remark this answers is not in the document]
Comment (tejaswiagarwal) [reply runtime/tchannel_client.go]: Done

File: runtime/tchannel_outbound_call.go [hunk 6]
```diff
@@ -89,15 +88,15 @@ func (c *tchannelOutboundCall) logFields(ctx context.Context) []zapcore.Field {
...
-		))
+	var logFields []zap.Field
+	if c.client != nil && c.client.contextExtractor != nil {
+		headers := map[string]string{}
+		for k, v := range c.reqHeaders {
+			headers[k] = v
+		}
+		ctx = WithEndpointRequestHeadersField(ctx, headers)
+		logFields = append(logFields, c.client.contextExtractor.ExtractLogFields(ctx)...)
+		ctx = WithLogFields(ctx, logFields...)
```
Comment (ChuntaoLu) [inline runtime/tchannel_outbound_call.go]: No need to put the log fields on to the context, the purpose of the method is to get log fields from context

File: runtime/tchannel_outbound_call.go [hunk 7]
```diff
@@ -89,15 +88,15 @@ func (c *tchannelOutboundCall) logFields(ctx context.Context) []zapcore.Field {
...
-		))
-	}
-	for k, v := range c.resHeaders {
-		fields = append(fields, zap.String(
-			fmt.Sprintf("%s-%s", logFieldClientResponseHeaderPrefix, k), v,
-		))
+	var logFields []zap.Field
+	if c.client != nil && c.client.contextExtractor != nil {
+		headers := map[string]string{}
+		for k, v := range c.reqHeaders {
```
Comment (ChuntaoLu) [inline runtime/tchannel_outbound_call.go]: This only includes request headers, what about response headers?
Also we need the prefix to differentiate the header keys between client and endpoint

File: test/endpoints/tchannel/baz/baz_simpleservice_method_call_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ChuntaoLu) [reply test/endpoints/tchannel/baz/baz_simpleservice_method_call_test.go]: I don't see any filter added to remove those from the logs, what happened here is that endpoint and client have the same header keys, and putting them on the context has an overwrite effect. We need the prefix to differentiate between those.

File: runtime/tchannel_outbound_call.go (hunk 6 above)
Comment (tejaswiagarwal) [reply runtime/tchannel_outbound_call.go]: Done!

File: examples/example-gateway/app.go [hunk 8]
```diff
@@ -62,14 +62,9 @@ func getRequestTags(ctx context.Context) map[string]string {
 func getRequestFields(ctx context.Context) []zap.Field {
 	var fields []zap.Field
 	headers := zanzibar.GetEndpointRequestHeadersFromCtx(ctx)
-	uuid, ok := headers["X-Uuid"]
```
Comment (ChuntaoLu) [inline examples/example-gateway/app.go]: We should not change this section, we deliberately avoided to have token as log fields here.

File: test/endpoints/tchannel/baz/baz_simpleservice_method_call_test.go [hunk 9]
```diff
@@ -178,9 +179,10 @@ func TestCallTChannelSuccessfulRequestOKResponse(t *testing.T) {
 		"endpointThriftMethod": "SimpleService::Call",
 		"endpointHandler":      "call",
 		"x-uuid":               "uuid",
-		"deviceversion":        "1.0",
-		"device":               "ios",
-		"regionname":           "sf",
+		"Deviceversion":        "1.0",
+		"Device":               "ios",
+		"Regionname":           "sf",
+		"x-token":              "token",
```
Comment (ChuntaoLu) [inline test/endpoints/tchannel/baz/baz_simpleservice_method_call_test.go]: The change in this PR should have no effects on the currently log fields except filtering.

Comment (tejaswiagarwal) [reply test/endpoints/tchannel/baz/baz_simpleservice_method_call_test.go]: Done, removed x-token and the other changes are for the casing.

Comment (ChuntaoLu) [inline test/endpoints/tchannel/baz/baz_simpleservice_method_call_test.go]: Token is still logged?

**Pull request merged**